### PR TITLE
draft of common retry and rate limiting decorators for api based modules

### PR DESCRIPTION
**Pull request by bcoca**
this should allow for a common way to add retry and rate limiting to new and existing API calls

### CODE REVIEW

File: lib/ansible/module_utils/api.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    """Retry decorator"""
+    def wrapper(f):
+        retry_count = 0
+        def retried(*args,**kargs):
+            if retries is not None:
+                ret = None
+                while True:
+                    retry_count += 1
+                    if retry_count >= retries:
+                        raise Exception("Retry limimit exceeded: %d" % retries)
```
Comment (jjshoe) [inline lib/ansible/module_utils/api.py]: s/limimit/limit/

File: lib/ansible/module_utils/api.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    arg_spec = (dict(
+        rate=dict(type='int'),
+        rate_limit=dict(type='int'),
+    ))
+    if spec:
+        arg_spec.update(spec)
+    return arg_spec
+
+def retry_argument_spec(spec=None):
+    """Creates an argument spec for working with rate limiting"""
```
Comment (jjshoe) [inline lib/ansible/module_utils/api.py]: "for working with retry limiting"

File: lib/ansible/module_utils/api.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+In order to use this module, include it as part of a custom
+module as shown below.
+
+** Note: The order of the import statements does matter. **
+
+from ansible.module_utils.basic import *
+from ansible.module_utils.api import *
+
+The eapi module provides the following common argument specs:
```
Comment (willthames) [inline lib/ansible/module_utils/api.py]: typo: api

**Comment by bw-jm** [comment]
@bcoca I have a few questions, I am running into tons of throttling issues with the AWS API as well eventual consistently problems.

- What is the state of building retries into all of the AWS modules?
- What do you think of doing this through decorators, something like this https://wiki.python.org/moin/PythonDecoratorLibrary#Retry ?

**Comment by bcoca** [comment]
- that is what they are here for
- They ARE decorators

**Comment by zentavr** [comment]
@bcoca - I wonder when this rate_limit feature will be available for AWS modules?

**Pull request merged**